### PR TITLE
Add riscv64 support

### PR DESCRIPTION
**Pull request by clefrks**
Adding support for riscv64 architecture.
Compiled and tested on VisionFive 2 board

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -364,11 +368,14 @@ ifeq ($(COMP),gcc)
...
 
-	ifeq ($(arch),$(filter $(arch),armv7 armv8))
+	ifeq ($(arch),$(filter $(arch),armv7 armv8 riscv64))
 		ifeq ($(OS),Android)
 			CXXFLAGS += -m$(bits)
 			LDFLAGS += -m$(bits)
 		endif
+		ifeq ($(ARCH),riscv64)
+                        CXXFLAGS += -latomic
+                endif
```
Comment (vondele) [inline src/Makefile:378]: spaces vs tabs here?

Comment (dav1312) [reply src/Makefile:378]: Currently the file already has some weird inconsistencies switching from tabs to spaces
https://github.com/official-stockfish/Stockfish/blob/79c5f3a69247b44ee6362a3d9236cd9bc048c5f5/src/Makefile#L381-L383

**Comment by clefrks** [comment]
`isn't this equivalent to arch=general-64`
in general-64, it does specify the bits, for riscv gnu compiler, There is no -m64 option. A riscv64-unknown-linux-gnu-gcc compiler emits 64-bit code by default, so there is no need for any option here. 
https://github.com/riscv-collab/riscv-gnu-toolchain/issues/670#issuecomment-656794799

`Is there any optimization (vectorization) that could be applied for risc-v?`
For VisionFive boards, we don't have any vector optimization yet.

attaching ./stockfish bench output
[bench.log](https://github.com/official-stockfish/Stockfish/files/9835276/bench.log)

File: src/Makefile (hunk 1 above)
Comment (clefrks) [reply src/Makefile]: My bad, updated the commit with tabs

**Comment by mkrcmarik** [comment]
Hi, I've been doing some riscv64 experiments, specifically on a visionfive 2 board with cpu 4x sfive-U74 @1.5GHz.
After compiling SF15.1, the "performance" measured in the bench is significantly below expectations (compared to raspberry pi 4x cortex-A72 @1.5GHz).

version SF15.1 on visionfive2 (bench 16 1 13 default depth mixed):

```
===========================
Total time (ms): 104780
Nodes searched : 3467381
Nodes/second : 33092
```


SF15.1 version on RPI4 (bench 16 1 13 default depth mixed)

```
===========================
Total time (ms): 12724
Nodes searched : 3467381
Nodes/second : 272507
```

Performance on armv8 is 8x higher than riscv64 at the same frequency.


I spent some time studying the possible causes and want to share some findings.

riscv64 processors have a specified set of ISA instructions, which has so-called extensions.
Not every riscv64 processor needs to include every extension for various reasons (e.g. ongoing riscv development).
For Stockfish, among others, the extension B - BMI (bit manipulation) and SIMD extension marked P or V (vector processing) are important, mainly for NNUE assessment
Not every compiler includes support for riscv64, or for every ISA extension (under development).

After many attempts, I came up with some positive results:
Using GCC since version 12.x (I'm using GCC 12.2 on Arch Linux) supports BMI instructions (eg cpop, ctz, clz), verified on the output of *.s files.
After editing the Makefile, it is possible to compile stockfish with BMI support, or POPCOUNT:

Edit Makefile:

```
### 2.2 Architecture specific

.
.
.


ifeq ($(ARCH),riscv64)
	arch = riscv64
	popcnt = yes

.
.
.



### ==========================================================================
### Section 3. Low-level Configuration
### ==========================================================================

### 3.1 Selecting compiler (default = gcc)

.
.
.

		ifeq ($(ARCH),riscv64)
			CXXFLAGS += -latomic -march=rv64imafdc_zba_zbb_zbc_zicsr
		endif
```

Added `popcnt = yes` and `-march=rv64imafdc_zba_zbb_zbc_zicsr` to support ISA extensions for CPUs (_zba _zbb and _zbc is a subset of B extensions)
The rv64imafdc_zba_zbb_zbc_zicsr chain is built based on ISA for sfive-u74

Output after command "compiler":

```
compiler

Compiled by g++ (GNUC) 12.2.0 on Linux
Compilation settings include: 64bit POPCNT
__VERSION__ macro expands to: 12.2.0
```


Bench results (bench 16 1 13 default depth mixed):

version SF15.1 on visionfive2 (-march=rv64imafdc_zba_zbb_zbc_zicsr):
```
===========================
Total time (ms) : 87309
Nodes searched : 3467381
Nodes/second : 39713
```

version SF15.1 on visionfive2 (popcnt = yes and -march=rv64imafdc_zba_zbb_zbc_zicsr):
```
===========================
Total time (ms) : 86367
Nodes searched : 3467381
Nodes/second : 40147
```


From the bench results, it is possible to see a partial improvement (20%) compared to SF15.1 without changes in the Makefile.


Meaning of SIMD:

After examining the source code, I found that the SIMD instructions are needed because of the NNUE processing.
We wanted to compare performance without the need for SIMD, so I turned off NNUE in the benchmark:


version SF15.1(popcount) on Visionfive2 (bench 16 1 13 default depth classical)
```
===========================
Total time (ms): 14735
Nodes searched : 4666594
Nodes/second : 316701
```


version SF15.1 on RPI4 (bench 16 1 13 default depth classical)
```
===========================
Total time (ms): 8281
Nodes searched : 4666594
Nodes/second : 563530
```

The benchmark shows that the main cause of low performance for visionfive2 is the absence of ISA extension for parallel data processing.
The solution may be to disable NNUE in the absence of SIMD, but this would require comparative tests on ELO in settings with and without NNUE.

**Comment by Ipmanchess** [comment]
Hi Michal ,are you interested to run with these two systems above for this bench list: [https://ipmanchess.yolasite.com/amd--intel-chess-bench-stockfish.php](url)

Thanks,Ipman.

**Comment by vondele** [comment]
> Hi, I've been doing some riscv64 experiments

Thanks.. also for x86-64 we have several variants to support more advanced ISA. So in principle one can target the more advanced riscv cpus specifically. I believe that you indeed picked the two most important aspects for performance, with bit manipulation and vector intrinsics. The performance gap is quite large though... which might make SF a good testcase. 

Switching of NNUE seems more like admitting defeat, I don't think that would be a good strategy.

**Pull request closed without merging**